### ISSUE TITLE
Idea for Reinforcement Learning from Device Feedback 

### ISSUE THREAD

**Issue by badeaa3** [badeaa3/asic-testing-fnal-lab-notebooks#56]
> Copied from smart-pix/asic-testing-fnal-lab-notebooks, original issue 45
> Originally opened by badeaa3 on 2024-11-19T15:31:03Z

Core of the idea -- use the physical chip during the training of the network or at a fine tuning step 

During training:
- Push the weights to the chip, do many injections, get the DNN result from chip
- Add term into loss function that encourages network to find weights that yield correct results on chip

**Comment by badeaa3** [comment]
> Copied from smart-pix/asic-testing-fnal-lab-notebooks, original issue 45
> Originally posted by badeaa3 on 2024-11-20T16:44:42Z

Model accuracy on the test dataset with no reweighting to a physical pT distribution:

- Full precision Tensorflow:
- Reduced precision QKeras: 0.7549742394504416 (From Jennet)
- C++: 
- RTL:
- On chip:

**Comment by badeaa3** [comment]
> Copied from smart-pix/asic-testing-fnal-lab-notebooks, original issue 45
> Originally posted by badeaa3 on 2024-12-05T02:37:03Z

to start on the ML training we need to do the following:
- compare the on chip results with the expected results for each input, rather than the RTL evaluated on the read back y-profile. This will give us an estimate of how much the “chip + testbench system” impacts the NN predictions. We need to quantify exactly what we aim to improve
- implement a spacely routine by copy-pasting the quantized training code from other ppl’s python scripts. run it without interfacing with the chip
- add into the function an intermediary step where the current training weights and inputs are sent to the chip. then we read back the chip predictions. add a term to the loss function that utilizes the readback. we could start with a MSE between the chip predictions (noisy) and the python output (idealized case), which the training should minimize. We will likely need to regulate/study it to make sure we dont end up at local (stupid) minima like all 0 weights

**Comment by badeaa3** [comment]
> Copied from smart-pix/asic-testing-fnal-lab-notebooks, original issue 45
> Originally posted by badeaa3 on 2024-12-05T21:39:12Z

Training code:
- https://github.com/smart-pix/smart-pixels-ml/blob/main/train_model.py
Model code:
- https://github.com/smart-pix/smart-pixels-ml/blob/main/models.py

We will copy these lines https://github.com/smart-pix/smart-pixels-ml/blob/main/train_model.py#L70-L131C2

**Comment by badeaa3** [comment]
> Copied from smart-pix/asic-testing-fnal-lab-notebooks, original issue 45
> Originally posted by bparpillon on 2024-12-06T03:03:34Z

**follow up  idea:**
We should  inject 100nA of current on _Ileak_ input to model a high radiation dose on every pixels.
Then we retrain the DNN and assess if the network can correct

**Comment by badeaa3** [comment]
> Copied from smart-pix/asic-testing-fnal-lab-notebooks, original issue 45
> Originally posted by badeaa3 on 2024-12-06T15:36:38Z

Set up environment:
```
source /mnt/local/gdg/tools/python/3.8.10/enable
source /asic/projects/C/CMS_PIX_28/testing/smartpixenv/bin/activate
```

Working directories
```
/asic/projects/C/CMS_PIX_28/abadea/directional-pixel-detectors/multiclassifier/SimplifiedTest.py
/asic/projects/C/CMS_PIX_28/abadea/working
```

- Cernbox link with datasets: https://cernbox.cern.ch/s/IAtFmNmA9qFGHPt
- Information about the different csv files: `https://github.com/GiuseppeDiGuglielmo/directional-pixel-detectors/commit/c92fab36e469ecc8f235c6db243d9889a78a04f3`
- Simplified running of model: https://github.com/GiuseppeDiGuglielmo/directional-pixel-detectors/blob/asic-flow/multiclassifier/SimplifiedTest.py#L78-L117
- C++ code https://github.com/GiuseppeDiGuglielmo/CMS_PIX_28_HLS/blob/main/hls4ml/d58w4a8_reduced/src/firmware/myproject.cpp
- Bit definitions in c++ https://github.com/GiuseppeDiGuglielmo/CMS_PIX_28_HLS/blob/main/hls4ml/d58w4a8_reduced/src/firmware/defines.h#L17
- RTL code https://github.com/GiuseppeDiGuglielmo/CMS_PIX_28_HLS/tree/main/rtl
- HLS4ml code https://github.com/GiuseppeDiGuglielmo/CMS_PIX_28_HLS/tree/main/hls4ml
- Notebook showing how to run training https://github.com/GiuseppeDiGuglielmo/directional-pixel-detectors/blob/asic-flow/multiclassifier/train.ipynb
- Weights coming out of hls4ml https://github.com/GiuseppeDiGuglielmo/CMS_PIX_28_HLS/tree/main/hls4ml/d58w4a8_reduced/src/firmware/weights
- hls4ml weight writer code https://github.com/fastmachinelearning/hls4ml/blob/main/hls4ml/writer/vivado_writer.py#L849
- Simplified test loading the qkeras model https://github.com/GiuseppeDiGuglielmo/directional-pixel-detectors/blob/asic-flow/multiclassifier/SimplifiedTest.py#L80-L94
- code to convert weights to binary /asic/projects/C/CMS_PIX_28/benjamin/verilog/workarea/cms28_smartpix_verification/PnR_cms28_smartpix_verification_A/tb/dnn/cms28_pix_dnn_wrapper_tb.py

**Comment by badeaa3** [comment]
> Copied from smart-pix/asic-testing-fnal-lab-notebooks, original issue 45
> Originally posted by badeaa3 on 2024-12-11T21:44:28Z

Attempting to make a single environment to run spacely and the training in:

- Must use python >= 3.10
```
INFO: pip is looking at multiple versions of fnal-libio to determine which version is compatible with other requirements. This could take a while.
ERROR: Package 'fnal-libio' requires a different Python: 3.8.10 not in '>=3.10'
```

Attempt 1:
- source /mnt/local/gdg/tools/python/3.10.14/enable

Attempted to compile with pointing to openssl and didn't work
```
(smartpixenv_p3.11.4) -bash-4.2$ which openssl
/usr/bin/openssl
(smartpixenv_p3.11.4) -bash-4.2$ openssl version -d
OPENSSLDIR: "/etc/pki/tls"
```

```
./configure --prefix=/asic/projects/C/CMS_PIX_28/testing/python/3.11.4 --enable-optimizations --with-lto --with-computed-gotos --with-system-ff --with-openssl=/etc/pki/tls
# from the configuration printout
checking for openssl/ssl.h in /etc/pki/tls... no
checking whether compiling and linking against OpenSSL works... no
checking for --with-openssl-rpath...
checking whether OpenSSL provides required ssl module APIs... no
checking whether OpenSSL provides required hashlib module APIs... no
```

**Working solution**

```
# working directory
cd /asic/projects/C/CMS_PIX_28/testing/tools/

# install openssl following https://docs.openiam.com/docs-4.2.1.3/appendix/2-openssl
wget https://www.openssl.org/source/openssl-1.1.1k.tar.gz
tar -xf openssl-1.1.1k.tar.gz
cd openssl-1.1.1k.tar
./config --prefix=/asic/projects/C/CMS_PIX_28/testing/tools/openssl --openssldir=/asic/projects/C/CMS_PIX_28/testing/tools/openssl shared zlib
make -j "$(nproc)"
make test
make install
cd ../

# point to tk libraries (or else later on problems after install)
export TCLTK_LIBS='-L/usr/lib -ltcl8.5 -ltk8.5'
export TCLTK_CFLAGS='-I/usr/include'

# install python 
wget https://www.python.org/ftp/python/3.11.11/Python-3.11.11.tgz
tar xvf Python-3.11.11.tgz
cd Python-3.11.11
# first without openssl or else it crashes
./configure --prefix=/asic/projects/C/CMS_PIX_28/testing/tools/python/3.11.11 --enable-optimizations --with-lto --with-computed-gotos --with-system-ffi
make -j "$(nproc)"
make altinstall
# then with openssl
./configure --prefix=/asic/projects/C/CMS_PIX_28/testing/tools/python/3.11.11 --enable-optimizations --with-lto --with-computed-gotos --with-system-ffi --with-openssl=/asic/projects/C/CMS_PIX_28/testing/tools/openssl
make -j "$(nproc)"
make altinstall

# install spacely
cd /asic/projects/C/CMS_PIX_28/abadea/working/CMSPIX28_DAQ/spacely/PySpacely
source SetupLinux.sh

# install ML requirements. Use a temporary directory or else run into space issue
cd /asic/projects/C/CMS_PIX_28/testing/tools
mkdir piptmp
TMPDIR=piptmp pip --cache-dir=/piptmp install -r requirements.txt

# tensorflow version matters
TMPDIR=/asic/projects/C/CMS_PIX_28/testing/tools/piptmp/ pip --cache-dir=/asic/projects/C/CMS_PIX_28/testing/tools/piptmp/ install tensorflow==2.13.0
```

Success!
**From now on:** `source /asic/projects/C/CMS_PIX_28/testing/tools/venvs/p3.11.11/bin/activate`

**Comment by badeaa3** [comment]
> Copied from smart-pix/asic-testing-fnal-lab-notebooks, original issue 45
> Originally posted by badeaa3 on 2024-12-12T04:35:44Z

Training within spacely successful! https://github.com/smart-pix/CMSPIX28Spacely/blob/badea_121124/CMSPIX28Spacely_Subroutines_B5_DNNTraining.py
```
# Spacely ready to take commands (see "help" for details)
<DBG> <2024-12-11 22:33:47> Unmonitored Rails: ['VTH2', 'VTH1', 'VTH0', 'VMC', 'SUPERPIX', 'INJ_1']
> ~r7
<DBG> <2024-12-11 22:33:52> Evaluating: ROUTINE_DNNTraining()
Model: "model"
_________________________________________________________________
 Layer (type)                Output Shape              Param #
=================================================================
 input1 (InputLayer)         [(None, 16)]              0

 dense1 (QDenseBatchnorm)    (None, 58)                1219

 relu1 (QActivation)         (None, 58)                0

 dense2 (QDense)             (None, 3)                 177

 linear (Activation)         (None, 3)                 0

=================================================================
Total params: 1396 (5.46 KB)
Trainable params: 1279 (5.00 KB)
Non-trainable params: 117 (472.00 Byte)
_________________________________________________________________
Interpreting Model
Topology:
Layer name: input1, layer type: InputLayer, input shapes: [[None, 16]], output shape: [None, 16]
Layer name: dense1, layer type: QDenseBatchnorm, input shapes: [[None, 16]], output shape: [None, 58]
Layer name: relu1, layer type: Activation, input shapes: [[None, 58]], output shape: [None, 58]
Layer name: dense2, layer type: QDense, input shapes: [[None, 58]], output shape: [None, 3]
Layer name: linear, layer type: Activation, input shapes: [[None, 3]], output shape: [None, 3]
Interpreting Model
Topology:
Layer name: input1, layer type: InputLayer, input shapes: [[None, 16]], output shape: [None, 16]
Layer name: dense1, layer type: QDenseBatchnorm, input shapes: [[None, 16]], output shape: [None, 58]
Layer name: relu1, layer type: Activation, input shapes: [[None, 58]], output shape: [None, 58]
Layer name: dense2, layer type: QDense, input shapes: [[None, 58]], output shape: [None, 3]
Layer name: linear, layer type: Activation, input shapes: [[None, 3]], output shape: [None, 3]
Creating HLS model
Writing HLS project
/asic/projects/C/CMS_PIX_28/testing/tools/venvs/p3.11.11/lib/python3.11/site-packages/keras/src/engine/training.py:3000: UserWarning: You are saving your model as an HDF5 file via `model.save()`. This file format is considered legacy. We recommend using instead the native Keras format, e.g. `model.save('my_model.keras')`.
  saving_api.save_model(
Done
     0      1      2
0  0.0 -0.375  0.375
(13982, 16) (13982,)
Epoch 1/10
11/11 [==============================] - 2s 39ms/step - loss: 0.5970 - sparse_categorical_accuracy: 0.7326 - val_loss: 0.6050 - val_sparse_categorical_accuracy: 0.7404
Epoch 2/10
11/11 [==============================] - 0s 5ms/step - loss: 0.5997 - sparse_categorical_accuracy: 0.7344 - val_loss: 0.6073 - val_sparse_categorical_accuracy: 0.7358
Epoch 3/10
11/11 [==============================] - 0s 4ms/step - loss: 0.5998 - sparse_categorical_accuracy: 0.7324 - val_loss: 0.6352 - val_sparse_categorical_accuracy: 0.7143
Epoch 4/10
11/11 [==============================] - 0s 4ms/step - loss: 0.6115 - sparse_categorical_accuracy: 0.7273 - val_loss: 0.6518 - val_sparse_categorical_accuracy: 0.6972
Epoch 5/10
11/11 [==============================] - 0s 4ms/step - loss: 0.6019 - sparse_categorical_accuracy: 0.7305 - val_loss: 0.6247 - val_sparse_categorical_accuracy: 0.7354
Epoch 6/10
11/11 [==============================] - 0s 4ms/step - loss: 0.5989 - sparse_categorical_accuracy: 0.7321 - val_loss: 0.6240 - val_sparse_categorical_accuracy: 0.7347
Epoch 7/10
11/11 [==============================] - 0s 4ms/step - loss: 0.6103 - sparse_categorical_accuracy: 0.7287 - val_loss: 0.6404 - val_sparse_categorical_accuracy: 0.7261
Epoch 8/10
11/11 [==============================] - 0s 4ms/step - loss: 0.5988 - sparse_categorical_accuracy: 0.7377 - val_loss: 0.6178 - val_sparse_categorical_accuracy: 0.7397
Epoch 9/10
11/11 [==============================] - 0s 4ms/step - loss: 0.5896 - sparse_categorical_accuracy: 0.7464 - val_loss: 0.6153 - val_sparse_categorical_accuracy: 0.7272
Epoch 10/10
11/11 [==============================] - 0s 4ms/step - loss: 0.6048 - sparse_categorical_accuracy: 0.7367 - val_loss: 0.6164 - val_sparse_categorical_accuracy: 0.7372
/asic/projects/C/CMS_PIX_28/testing/tools/venvs/p3.11.11/lib/python3.11/site-packages/keras/src/engine/training.py:3000: UserWarning: You are saving your model as an HDF5 file via `model.save()`. This file format is considered legacy. We recommend using instead the native Keras format, e.g. `model.save('my_model.keras')`.
  saving_api.save_model(
Save: model.h5
437/437 [==============================] - 1s 1ms/step - loss: 0.6016 - sparse_categorical_accuracy: 0.7429
Test loss: 0.6015514135360718
Test accuracy: 0.7428836822509766
<INF> <2024-12-11 22:34:00> This Routine took: 0:00:08
```
